cmd/snap/cmd_buy.go

```diff
@@ -117,33 +110,28 @@ Once completed, return here and run 'snap buy %s' again.`), snap.Name)
 		return err
 	}
 
-	reader := bufio.NewReader(nil)
-	reader.Reset(Stdin)
+	fmt.Fprintf(Stdout, i18n.G(`Please re-enter your Ubuntu One password to purchase '%s' from '%s' for %s. Press ctrl-c to cancel.`),
+		snap.Name, snap.Developer, formatPrice(opts.Price, opts.Currency))
+	fmt.Fprint(Stdout, "\n")
 
-	fmt.Fprintf(Stdout, i18n.G("Do you want to buy %q from %q for %s? (Y/n): "), snap.Name,
-		snap.Developer, formatPrice(opts.Price, opts.Currency))
-
-	response, _, err := reader.ReadLine()
+	user, err := client.ReadAuthData()
```

Comment:
We shouldn't have to re-read this data here. It's internal to the client, and it should be responsible for re-sending and re-authenticating whenever necessary, even on re-logins. Re-authenticating from the ground up as a new user is likely the reason why you ended up with several independent backing users on snapd.